resources/execute_event_handler.rb

```diff
@@ -0,0 +1,182 @@
+            byos_substack_outputs.merge!({ 'Outputs' => { substack_output['Key'] => substack_output['Value'] } })
+          end
+          ::File.write(source_byos_substack_outputs, byos_substack_outputs.to_json(:only))
+        end
+      end
+      FileUtils.cp(source_byos_substack_outputs, target_byos_substack_outputs)
+    end
+
+    source_handler_env = '/opt/parallelcluster/shared/handler_env.json'
+    if ::File.exist?(source_handler_env)
+      Chef::Log.info("Found handler environment file (#{source_handler_env})")
+      env = JSON.load_file(source_handler_env)
+      Chef::Log.debug("Loaded handler environment #{env}")
+    else
+      Chef::Log.info("No handler environment file found, building it")
+      env = build_static_env(target_cluster_config, target_launch_templates, target_byos_substack_outputs)
+
+      Chef::Log.info("Dumping handler environment to file (#{source_handler_env})")
+      ::File.write(source_handler_env, env.to_json(:only))
+    end
+
+    env.merge!(build_dynamic_env)
+    env
+  end
+
+  def build_dynamic_env
+    Chef::Log.info("Building dynamic handler environment")
+    env = {}
+
+    # PCLUSTER_EC2_INSTANCE_TYPE
+    env.merge!(build_hash_from_node('PCLUSTER_EC2_INSTANCE_TYPE', :ec2, :instance_type))
+
+    case node['cluster']['node_type']
+    when 'ComputeFleet'
+      # PCLUSTER_QUEUE_NAME
+      env.merge!(build_hash_from_node('PCLUSTER_QUEUE_NAME', :cluster, :scheduler_queue_name))
+
+      # PCLUSTER_COMPUTE_RESOURCE_NAME
+      env.merge!(build_hash_from_node('PCLUSTER_COMPUTE_RESOURCE_NAME', :cluster, :scheduler_compute_resource_name))
+    end
+
+    env
+  end
+
+  def build_static_env(target_cluster_config, target_launch_templates, target_byos_substack_outputs)
+    Chef::Log.info("Building static handler environment")
+    env = {}
+
+    # PCLUSTER_CLUSTER_CONFIG
+    if ::File.exist?(target_cluster_config)
+      env.merge!({ 'PCLUSTER_CLUSTER_CONFIG' => target_cluster_config })
+    end
+
+    # PCLUSTER_LAUNCH_TEMPLATES
+    if ::File.exist?(target_launch_templates)
+      env.merge!({ 'PCLUSTER_LAUNCH_TEMPLATES' => target_launch_templates })
+    end
+
+    # PCLUSTER_CLUSTER_NAME
+    env.merge!(build_hash_from_node('PCLUSTER_CLUSTER_NAME', :cluster, :stack_name))
+
+    # PCLUSTER_CFN_STACK_ARN
+    env.merge!(build_hash_from_node('PCLUSTER_CFN_STACK_ARN', :cluster, :stack_arn))
+
+    # PCLUSTER_BYOS_CFN_SUBSTACK_ARN
+    env.merge!(build_hash_from_node('PCLUSTER_BYOS_CFN_SUBSTACK_ARN', :cluster, :byos_substack_arn))
+
+    # PCLUSTER_BYOS_CFN_SUBSTACK_OUTPUTS
+    if ::File.exist?(target_byos_substack_outputs)
+      env.merge!({ 'PCLUSTER_BYOS_CFN_SUBSTACK_OUTPUTS' => target_byos_substack_outputs })
+    end
+
+    # PCLUSTER_SHARED_SCHEDULER_DIR
+    env.merge!({ 'PCLUSTER_SHARED_SCHEDULER_DIR' => '/opt/parallelcluster/shared/byos' })
+
+    # PCLUSTER_LOCAL_SCHEDULER_DIR
+    env.merge!({ 'PCLUSTER_LOCAL_SCHEDULER_DIR' => '/opt/parallelclusterbyos' })
```

Comment:
missing `/`. might be worth defining all this dirs in a single place (attributes?)

Reply:
done

Reply:
tested removing the DescribeStacks policy and got the following

```
[2021-10-06T09:03:22+00:00] INFO: Executing describe-stack on byos substack (arn:aws:cloudformation:eu-south-1:346106133209:stack/byos2/c729ce70-252b-11ec-99af-0ae7da62c0e4)


    ================================================================================
    Error executing action `run` on resource 'execute_event_handler[HeadInitEnvironment]'
    ================================================================================

    RuntimeError
    ------------
    Unable to execute describe-stack on byos substack (arn:aws:cloudformation:eu-south-1:346106133209:stack/byos2/c729ce70-252b-11ec-99af-0ae7da62c0e4)
     ---- STDERR for HeadInitEnvironment Event ----
    An error occurred (AccessDenied) when calling the DescribeStacks operation: User: arn:aws:sts::346106133209:assumed-role/byos2-HeadNodeIamResourcesRoleHeadNodeA2902647-LP3ANJMMOIQ8/i-00294aae32e63294e is not authorized to perform: cloudformation:DescribeStacks on resource: arn:aws:cloudformation:eu-south-1:346106133209:stack/byos2/c729ce70-252b-11ec-99af-0ae7da62c0e4
    ---- End STDERR for HeadInitEnvironment Event ----
```